xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,439 @@
+            return;
+          }
+          if (update.listener() == null) {
+            // Ignore updates not for servers.
+            return;
+          }
+          filterChains = update.listener().getFilterChains();
+          defaultFilterChain = update.listener().getDefaultFilterChain();
+          updateSelector();
+        }
+      });
+    }
+
+    @Override
+    public void onResourceDoesNotExist(final String resourceName) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          if (stopped) {
+            return;
+          }
+          handleConfigNotFound(null);
+          listener.onNotServing(
+              Status.UNAVAILABLE.withDescription(
+                  "Listener " + resourceName + " unavailable").asException());
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Status error) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          if (stopped) {
+            return;
+          }
+          logger.log(Level.FINE, "Transient error from XdsClient: {0}", error);
+          if (isPermanentError(error)) {
+            handleConfigNotFound(null);
+          }
+          if (!isServing) {
+            listener.onNotServing(error.asException());
+          }
+        }
+      });
+    }
+
+    private void shutdown() {
+      stopped = true;
+      logger.log(Level.FINE, "Stop watching LDS resource {0}", resourceName);
+      xdsClient.cancelLdsResourceWatch(resourceName, this);
+      List<SslContextProviderSupplier> toRelease = collectSslContextProviderSuppliers();
+      filterChainSelectorRef.set(FilterChainSelector.NO_FILTER_CHAIN);
+      for (SslContextProviderSupplier s: toRelease) {
+        s.close();
+      }
+    }
+
+    private List<SslContextProviderSupplier> collectSslContextProviderSuppliers() {
+      List<SslContextProviderSupplier> toRelease = new ArrayList<>();
+      FilterChainSelector selector = filterChainSelectorRef.get();
+      if (selector != null) {
+        for (FilterChain f: selector.getFilterChains()) {
+          if (f.getSslContextProviderSupplier() != null) {
+            toRelease.add(f.getSslContextProviderSupplier());
+          }
+        }
+        SslContextProviderSupplier defaultSupplier =
+                selector.getDefaultSslContextProviderSupplier();
+        if (defaultSupplier != null) {
+          toRelease.add(defaultSupplier);
+        }
+      }
+      return toRelease;
+    }
+
+    private void updateSelector() {
+      List<SslContextProviderSupplier> toRelease = collectSslContextProviderSuppliers();
+      FilterChainSelector selector = new FilterChainSelector(
+          Collections.unmodifiableList(filterChains),
+          defaultFilterChain == null ? null : defaultFilterChain.getSslContextProviderSupplier());
+      filterChainSelectorRef.set(selector);
+      for (SslContextProviderSupplier s: toRelease) {
+        s.close();
+      }
```

Comment:
You can refactor lines 399-406 to create a method, say:

```
void closeAndReplace(FilterChainSelector selector) {
      List<SslContextProviderSupplier> toRelease = collectSslContextProviderSuppliers();
      filterChainSelectorRef.set(selector);
      for (SslContextProviderSupplier s: toRelease) {
        s.close();
      }
}
``` 

And call it from here and also from `DiscoveryState.shutdown` as

```
closeAndReplace(FilterChainSelector.NO_FILTER_CHAIN);
```